### PR TITLE
implement `gt ship`

### PR DESCRIPTION
**Pull request by charlierudolph**

### CODE REVIEW

File: features/git-town-ship/current_branch/on_feature_branch/without_open_changes/in_subfolder.feature [hunk 1]
```diff
@@ -39,18 +39,18 @@ Feature: git town-ship: shipping the current feature branch from a subfolder
...
-      |         | git revert <%= sha 'feature done' %>           |
-      |         | git push                                       |
-      |         | git checkout feature                           |
-      | feature | git checkout main                              |
-      | main    | git checkout feature                           |
-      | <none>  | cd <%= git_folder "new_folder" %>              |
+      | BRANCH  | COMMAND                                                        |
+      | <none>  | cd <%= git_root_folder %>                                      |
+      | main    | git branch feature <%= sha 'feature commit' %>                 |
+      |         | git push origin <%= sha 'feature commit' %>:refs/heads/feature |
```
Comment (kevgo) [inline features/git-town-ship/current_branch/on_feature_branch/without_open_changes/in_subfolder.feature]: Should we add the `-u` back to make the remote branch track the local branch? I think that's useful to have.

File: lib/git/status.go [hunk 2]
```diff
@@ -4,9 +4,16 @@ import (
 	"fmt"
 	"os"
 
+	"github.com/Originate/git-town/lib/config"
 	"github.com/Originate/git-town/lib/util"
 )
 
+func EnsureDoesNotHaveUncommitedChanges(message string) {
+	if HasOpenChanges() {
```
Comment (kevgo) [inline lib/git/status.go]: Let's standardize on either "uncommitted changes" or "open changes". The two methods on this line and the one above use different terminology for the same thing.

File: lib/steps/commit_open_changes_step.go [hunk 3]
```diff
@@ -22,10 +22,18 @@ func (step CommitOpenChangesStep) CreateUndoStep() Step {
 	return ResetToShaStep{Sha: git.GetBranchSha(branchName)}
 }
 
+func (step CommitOpenChangesStep) GetAutomaticAbortErrorMessage() string {
+	return ""
```
Comment (kevgo) [inline lib/steps/commit_open_changes_step.go]: What are functions like this one that don't do anything for?

File: lib/steps/delete_parent_branch_step.go [hunk 4]
```diff
@@ -23,7 +23,15 @@ func (step DeleteParentBranchStep) CreateUndoStep() Step {
 	}
 }
 
+func (step DeleteParentBranchStep) GetAutomaticAbortErrorMessage() string {
```
Comment (kevgo) [inline lib/steps/delete_parent_branch_step.go]: Hmm, there must be a better way to do this that avoids all this duplication. Traditionally I would have moved this method into the base class and then override in subclasses as needed. Go achieves the same thing with composition. [Here](http://stackoverflow.com/questions/31997753/golang-reuse-a-function-across-multiple-structs-to-satisfy-an-interface) is an idea on how this could be done.

File: lib/steps/fetch_step.go [hunk 5]
```diff
@@ -6,6 +6,8 @@ import (
 	"github.com/Originate/git-town/lib/script"
 )
 
+var fetched bool
```
Comment (kevgo) [inline lib/steps/fetch_step.go]: Ideally we do this without mutable state in this file. Maybe filter out repeated `FetchStep` instances from the steplist? This can be done later.

File: lib/steps/load_state.go [hunk 6]
```diff
@@ -5,18 +5,34 @@ import (
...
 )
 
-func loadState(command string) RunState {
+func hasSavedState(command string) bool {
 	filename := getRunResultFilename(command)
-	content, err := ioutil.ReadFile(filename)
-	if err != nil {
-		log.Fatal(err)
+	_, err := os.Stat(filename)
+	return !os.IsNotExist(err)
```
Comment (kevgo) [inline lib/steps/load_state.go]: What if there is another error? Then this function wrongfully returns `true`.

File: lib/steps/delete_parent_branch_step.go (hunk 4 above)
Comment (charlierudolph) [inline lib/steps/delete_parent_branch_step.go]: Looks awesome. That will reduce a huge amount of duplcation

File: lib/git/status.go (hunk 2 above)
Comment (charlierudolph) [inline lib/git/status.go]: Sounds good. Already using open changes everywhere else so will go with that

File: features/git-town-ship/current_branch/on_feature_branch/without_open_changes/in_subfolder.feature (hunk 1 above)
Comment (charlierudolph) [inline features/git-town-ship/current_branch/on_feature_branch/without_open_changes/in_subfolder.feature]: Sounds good.

File: cmd/ship.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	} else {
+		result.TargetBranch = args[0]
+	}
+	if config.HasRemote("origin") {
+		steps.FetchStep{}.Run()
+	}
+	if result.TargetBranch != result.InitialBranch {
+		git.EnsureHasBranch(result.TargetBranch)
+	}
+	config.EnsureIsFeatureBranch(result.TargetBranch, "Only feature branches can be shipped.")
```
Comment (charlierudolph) [inline cmd/ship.go]: I'd create an issue for this.

File: cmd/ship.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		return validateMaxArgs(args, 1)
+	},
+}
+
+func checkShipPreconditions(args []string) (result ShipConfig) {
+	result.InitialBranch = git.GetCurrentBranchName()
+	if len(args) == 0 {
+		result.TargetBranch = result.InitialBranch
+		git.EnsureDoesNotHaveUncommitedChanges("Did you mean to commit them before shipping?")
```
Comment (charlierudolph) [inline cmd/ship.go]: If you supply a branch name that is different then what you are currently on, we stash the changes. I'm actually surprised this is passing. Looks like we are missing a test case where you supply the current branch name

File: lib/steps/fetch_step.go (hunk 5 above)
Comment (charlierudolph) [inline lib/steps/fetch_step.go]: I don't think this is needed as I updated fetch to always run in preconditions. I'll remove this

File: lib/steps/delete_parent_branch_step.go (hunk 4 above)
Comment (charlierudolph) [inline lib/steps/delete_parent_branch_step.go]: Done. I'll add more composition to reduce some other duplication in another PR.

File: lib/git/status.go (hunk 2 above)
Comment (charlierudolph) [inline lib/git/status.go]: done

File: cmd/ship.go (hunk 8 above)
Comment (charlierudolph) [inline cmd/ship.go]: Created an issue for the missing test

File: lib/steps/fetch_step.go (hunk 5 above)
Comment (charlierudolph) [inline lib/steps/fetch_step.go]: removed

File: cmd/ship.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	if config.HasRemote("origin") {
+		steps.FetchStep{}.Run()
+	}
+	if result.TargetBranch != result.InitialBranch {
+		git.EnsureHasBranch(result.TargetBranch)
+	}
+	config.EnsureIsFeatureBranch(result.TargetBranch, "Only feature branches can be shipped.")
+	prompt.EnsureKnowsParentBranches([]string{result.TargetBranch})
+	if config.GetParentBranch(result.TargetBranch) != config.GetMainBranch() {
```
Comment (charlierudolph) [inline cmd/ship.go]: done

File: lib/steps/load_state.go (hunk 6 above)
Comment (charlierudolph) [inline lib/steps/load_state.go]: fixed

File: features/git-town-ship/current_branch/on_feature_branch/without_open_changes/in_subfolder.feature (hunk 1 above)
Comment (charlierudolph) [inline features/git-town-ship/current_branch/on_feature_branch/without_open_changes/in_subfolder.feature]: fixed

**Comment by charlierudolph** [comment]
all comments addressed

File: lib/prompt/squash_commit_author.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	} else {
+		util.PrintError("Invalid author number")
+		return ""
+	}
+}
+
+func printNumberedAuthors(authors []Author) {
+	boldFmt := color.New(color.Bold)
+	for index, author := range authors {
+		stat := author.NumberOfCommits + " commit"
```
Comment (kevgo) [inline lib/prompt/squash_commit_author.go]: Extract a `pluralize` method similar to the one in [Rails](https://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize)?

File: lib/steps/create_and_checkout_branch_step.go [hunk 11]
```diff
@@ -6,6 +6,7 @@ import (
 )
 
 type CreateAndCheckoutBranchStep struct {
+	NoAutomaticAbort
```
Comment (kevgo) [inline lib/steps/create_and_checkout_branch_step.go]: There is a method called `ShouldAutomaticallyAbortOnError`, so I would expect this to be named `NoAutomaticAbortOnError`.

File: lib/steps/discard_open_changes_step.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"github.com/Originate/git-town/lib/script"
+)
+
+type DiscardOpenChangesStep struct {
+	NoAutomaticAbort
+}
+
+func (step DiscardOpenChangesStep) CreateAbortStep() Step {
```
Comment (kevgo) [inline lib/steps/discard_open_changes_step.go]: These repetitive methods are another candidate for DRYing up using composition in another PR!

File: lib/steps/load_state.go [hunk 13]
```diff
@@ -5,18 +5,41 @@ import (
...
 )
 
-func loadState(command string) RunState {
+func hasSavedState(command string) bool {
 	filename := getRunResultFilename(command)
-	content, err := ioutil.ReadFile(filename)
+	_, err := os.Stat(filename)
 	if err != nil {
-		log.Fatal(err)
+		if os.IsNotExist(err) {
```
Comment (kevgo) [inline lib/steps/load_state.go]: Rewrite without if-clause nesting and without else:

```go
if err == nil {
  return true
}
if os.IsNotExist(err) {
  return false
}
log.Fatal("Error getting stat for run result file:", err)

File: lib/steps/discard_open_changes_step.go (hunk 12 above)
Comment (charlierudolph) [inline lib/steps/discard_open_changes_step.go]: Already have an issue for that

File: lib/steps/create_and_checkout_branch_step.go (hunk 11 above)
Comment (charlierudolph) [inline lib/steps/create_and_checkout_branch_step.go]: sounds good

File: lib/steps/load_state.go (hunk 13 above)
Comment (charlierudolph) [inline lib/steps/load_state.go]: go doesn't let that compile. It errors with: `lib/steps/load_state.go:21: missing return at end of function`

Comment (kevgo) [inline lib/steps/load_state.go]: Oh. Okay, please add the return at the end. `golint` does not like if-else clauses that all contain `return` statements. It suggests to remove the last `else` statement and unindent the block inside it.

File: lib/prompt/squash_commit_author.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+func parseAuthorNumber(userInput string, authors []Author) string {
+	index, err := strconv.Atoi(userInput)
+	if err != nil {
+		log.Fatal("Error parsing string to integer: ", err)
+	}
+	if index >= 1 && index <= len(authors) {
+		return authors[index-1].NameAndEmail
+	} else {
+		util.PrintError("Invalid author number")
+		return ""
```
Comment (charlierudolph) [inline lib/prompt/squash_commit_author.go]: Can I do this in another PR? This is how its done in the parent branch prompt currently and I'd like to update those in unison.

Comment (kevgo) [inline lib/prompt/squash_commit_author.go]: You can do anything in another PR  :)

File: features/git-town-ship/current_branch/on_feature_branch/on_parent_branch.feature [hunk 15]
```diff
@@ -42,11 +42,11 @@ Feature: git town-ship: shipping a parent branch
 
 
   Scenario: undo
-    When I run `git town-ship --undo`
+    When I run `gt ship --undo`
     Then it runs the commands
       | BRANCH         | COMMAND                                                      |
       | main           | git branch parent-feature <%= sha 'parent feature commit' %> |
-      |                | git revert <%= sha 'parent feature done' %>                  |
+      |                | git revert HEAD                                              |
```
Comment (kevgo) [inline features/git-town-ship/current_branch/on_feature_branch/on_parent_branch.feature]: Unless I am misunderstanding you, we should be able to determine the SHA of the commit we want to revert to __before__ we run the squash-merge step, no?

Comment (charlierudolph) [inline features/git-town-ship/current_branch/on_feature_branch/on_parent_branch.feature]: Tried doing the following and it gave a different sha:

* squash branch
* commit
* get sha
* reset to previous commit

The sha that resulted in ended up being different from the sha of the actual commit. I believe the sha includes the timestamp of the commit.

File: lib/steps/load_state.go (hunk 13 above)
Comment (charlierudolph) [inline lib/steps/load_state.go]: I think the current implementation is sufficient then. Please take another look.

**Comment by charlierudolph** [comment]
@kevgo all comments addressed

File: lib/steps/load_state.go (hunk 13 above)
Comment (kevgo) [inline lib/steps/load_state.go]: This is okay for now. Let's see what `golint` thinks about it. Btw is your editor running Golint when you save the file? This is super useful in my experience.
